### PR TITLE
Linting, Testing, Coverage, and CI

### PR DESCRIPTION
**Pull request by ChrisAlderson**
Not sure if you are happy with this one because it breaks about everything for all the implementing packages.

 - Linting with [`eslint`](https://github.com/standard/eslint-config-standard) with a configuration based on [StandardJS](https://github.com/standard/standard)
 - Moved the `base.js` tests from `tape` to [`chai`](https://github.com/chaijs/chai) and [`mocha`](https://github.com/mochajs/mocha)
 - Added `mocha` tests for the `butter-provider` package
 - Added code coverage with [`nyc`](https://github.com/istanbuljs/nyc)
 - Added CI with `travis`

The tests defined in `basic.js` can still be executed by the implementations of the `butter-provider`, but since the directory holding the tests files has been changed from `tests` to `test` you probably need to make changes in the sub packages. The change of directories is because of mocha.

### CODE REVIEW

**Comment by xaiki** [comment]
do we need to move the test directory ?
it's a bit of a hassle.

File: index.js [hunk 1]
```diff
@@ -58,7 +59,7 @@ class Provider {
 
     const { argTypes, defaults } = this.config
     Object.keys(argTypes).map(k => {
-      if (!argTypes || !argTypes[k]) {
+      if (!parsed || !parsed[k]) {
```
Comment (xaiki) [inline index.js]: this is needed anyway, can you PR it out ?

**Comment by xaiki** [comment]
it looks very good.
i'm ok to make one provider overhaul, but i'd like it that we don't have to change all the provider yet another time. can we find a test entrypoint that'd be stable so that we *never* have to change all the providers again ?

File: index.js (hunk 1 above)
Comment (ChrisAlderson) [inline index.js]: Well here you want to check if all the `args` in the `config` are parsed, right? So what should be done is:
 - parse the `argString` to an object
 - merge the default args from `config` with the parsed `argString`
 - loop through the needed arguments
    - check if a needed argument exists
       - if it does 👍 
       - if it does not 👎

**Comment by ChrisAlderson** [comment]
Looks like the test directory didn't need to be changed after all, so I changed it back. 
Also reverted something from my previous PR, where I changed `config.args` to `config.argTypes`

File: index.js [hunk 2]
```diff
@@ -16,7 +16,7 @@ const defaultArgs = {
 }
 
 const defaultConfig = {
-  argTypes: {},
+  args: {},
```
Comment (xaiki) [inline index.js]: this patch needs to be dropped

**Comment by xaiki** [comment]
6482bf5 needs to be dropped.

File: index.js [hunk 3]
```diff
@@ -57,13 +58,15 @@ class Provider {
       : undefined
 
     const { args, defaults } = this.config
```
Comment (xaiki) [inline index.js]: argTypes

File: index.js [hunk 4]
```diff
@@ -57,13 +58,15 @@ class Provider {
...
+    const argsObject = Object.assign({}, defaults, parsed)
```
Comment (xaiki) [inline index.js]: this can become args

File: index.js [hunk 5]
```diff
@@ -74,8 +77,9 @@ class Provider {
     if (tokenize[1]) {
       tokenize[1].split('&').map(v => {
         const [ key, value ] = v.split('=')
+        const type = this.config.args[key]
 
-        args[key] = this._parseArgForType(key, value)
+        args[key] = this._parseArgForType(type, value)
```
Comment (xaiki) [inline index.js]: this needs to get rebased on the fix-es6 branch

File: package-lock.json [hunk 6]
```diff
@@ -0,0 +1,4069 @@
+{
```
Comment (xaiki) [inline package-lock.json]: is this supposed to be in ?

File: package.json [hunk 7]
```diff
@@ -4,7 +4,9 @@
   "description": "A base class for Butter providers",
   "main": "index.js",
   "scripts": {
-    "test": "echo \"Error: no test specified\" && exit 1"
+    "coveralls": "cat ./coverage/lcov.info | node ./node_modules/coveralls/bin/coveralls",
+    "lint": "./node_modules/.bin/eslint tests/** index.js",
+    "test": "./node_modules/.bin/nyc --reporter=lcov --reporter=text ./node_modules/.bin/mocha tests/*.spec.*"
```
Comment (xaiki) [inline package.json]: shouldn't this be `test/` ?

File: index.js (hunk 4 above)
Comment (ChrisAlderson) [inline index.js]: fixed

File: index.js (hunk 3 above)
Comment (ChrisAlderson) [inline index.js]: fixed

File: package.json (hunk 7 above)
Comment (ChrisAlderson) [inline package.json]: No I changed the directory back in `fbdc150`.

**Comment by xaiki** [comment]
need to remove the lock file commited in 331e75540891d0d285515e410e777f9adeb6030c, if you're unconfortable doing that i can merge and rewrite history later.

**Pull request merged**